### ISSUE TITLE
Define behavior of pressing enter in inline editing host 

### ISSUE THREAD

**Issue by foolip** [w3c/editing#430]
There are tentative tests for this:
https://github.com/web-platform-tests/wpt/blob/master/editing/other/insertparagraph-in-inline-editing-host.tentative.html
https://wpt.fyi/results/editing/other?label=experimental&label=master&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2023-webcompat

The tests are part of the Interop 2023 Web Compat focus area, but there's no spec for it.

If the behavior can be written down in some specification (whatever its standards track status) managed by the Editing WG, it will be much easier to implement something in Chromium that we can be confident matches the consensus of the WG.

(To reverse engineer the exact behavior from the tests isn't generally possible unless tests cover all possibilities.)

**Comment by foolip** [comment]
@chrishtr can you help get this on the Editing WG agenda?

It would be good to have an editing expert read the test and summarize in simpler terms what the important aspects are. I think it has to do with inserting a different element to represent the new line/paragraph depending on what the parent element is, to not end up with `<span><div></div><span>` for example.

**Comment by masayuki-nakano** [comment]
> I think it has to do with inserting a different element to represent the new line/paragraph depending on what the parent element is, to not end up with `<span><div></div><span>` for example.

I intended so. 

I think `insertParagraph` should not insert a default paragraph separator `<div>` or `<p>`, if there is no proper editable ancestor at insertion point. In the case, the command should work as `insertLineBreak` (it seems that `\n` should be inserted if it's preserved by `white-space`, other wise, `<br>`).

**Comment by foolip** [comment]
Thanks for explaining @masayuki-nakano! How should "proper editable ancestor" be defined here?

**Comment by masayuki-nakano** [comment]
I guess that these steps may roughly match with existing `insertParagraph` command behavior of the browsers:
1. Let _inclusive ancestor block_ be null
2. Let _container_ be focus node of selection (after deleting selected content)
3. If _container_ can not have the new default paragraph as a child
 3-1. If _container_ creates a context, e.g., `<caption>`, `<body>`, etc, break this loop
 3-2. If parent node of _container_ is not editable, clear _container_ and break this loop
 3-3. If parent node of _container_ is a `<div>`, `<p>`, `<hn>`, `<li>`, `<dt>` or `<dd>`, break this loop
 3-4. Set _container_ to the parent, then back to the step 3
5. If _container_ is not set, fallback to `insertLineBreak` and return
6. If _container_ is an element creating a context (i.e., broken the loop at 3-1), put the default paragraph in it or insert a line break instead if it cannot have the new paragraph, then, return
7. Duplicate _container_ and put the clone after it (and delete unnecessary content)

**Comment by chrishtr** [comment]
> 3\. 3-1. If _container_ creates a context, e.g., `<caption>`, `<body>`, etc, break this loop

Hi @masayuki-nakano, could you clarify what you mean by "context" here?

> 3\. If _container_ can not have the new default paragraph as a child

Could you say what conditions lead to "can not have the new default paragraph as a child"?  Does this mean "if the container is inline"? If so, would that be determined by tag name or actual inline style?

> 3\. 3-3. If parent node of _container_ is a `<div>`, `<p>`, `<hn>`, `<li>`, `<dt>` or `<dd>`, break this loop

Could you say more about why this specific list of elements?

**Comment by masayuki-nakano** [comment]
> > 3. 3-1. If _container_ creates a context, e.g., `<caption>`, `<body>`, etc, break this loop
> 
> Hi @masayuki-nakano, could you clarify what you mean by "context" here?

I meant that it creates a things which are enclosed in a group of content, it should not be duplicated like `<caption>`, etc. (I'm not sure the good term to explain such elements.)

> > 3. If _container_ can not have the new default paragraph as a child
> 
> Could you say what conditions lead to "can not have the new default paragraph as a child"? Does this mean "if the container is inline"? If so, would that be determined by tag name or actual inline style?

I'm thinking about the relation between parent element and child element (new default paragraph) from HTML spec point of view.

> > 3. 3-3. If parent node of _container_ is a `<div>`, `<p>`, `<hn>`, `<li>`, `<dt>` or `<dd>`, break this loop
> 
> Could you say more about why this specific list of elements?

They won't have new default paragraphs in any browsers, I think. They will become the left node, and browsers put the right node for it.

**Comment by chrishtr** [comment]
I discussed it within Chromium and we think that the algorithm can be simplified to insert a simple line break (`<br>`) if one of the following holds:

* There is no enclosing block of the selection position
* The enclosing block is one of these:
    * A [Phrasing Content](https://developer.mozilla.org/en-US/docs/Web/HTML/Content_categories#phrasing_content) element
    * A table cell or table element
* An `<hr>` element

This is implemented in Chromium [here](https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:third_party/blink/renderer/core/editing/commands/insert_paragraph_separator_command.cc;drc=a742c65cc620d2b7b262b5514540ebdb510a7121;bpv=1;bpt=1;l=238?q=insert_paragraph_separator_command).

The part about phrasing content was added recently, and has not yet been successfully shipped to the stable channel. It's based on an [equivalent change](https://github.com/WebKit/WebKit/commit/37c9962af2e3af8339a71539efdcd40c813d6bd6#diff-62b93c3eda712d14d0be1084c4bd28bc49bf5e81af3b9ca07e552065324dc5f4) to WebKit.

**Comment by masayuki-nakano** [comment]
@chrishtr What is the "enclosing block" definition? Which one if `<div><span contenteditable>{}<br><span></div>`? Sounds like the non-editable `<div>` in this case.

**Comment by chrishtr** [comment]
> @chrishtr What is the "enclosing block" definition? Which one if `<div><span contenteditable>{}<br><span></div>`? Sounds like the non-editable `<div>` in this case.

It's the highest editable ancestor element in the DOM which is a block. If the `{}` in your example is the caret position, then there would be no enclosing block, because the `span` element is not a block, and no ancestor of the `span` is editable.

**Comment by johanneswilm** [comment]
TPAC 2023
RESOLUTION: Simon will document what happens in the three browser engines and check existing wpt tests. Then pick one of the behaviors.

**Comment by foolip** [comment]
Was there some further update on this? Is Simon @zcorpan or another Simon?

**Comment by zcorpan** [comment]
It was probably me! Apologies for the delay.

The tests referenced in OP pass now! So it seems we already have interoperable behavior and the spec needs updating to reflect what is implemented.
